File: benchmarks/baselines/typescript/v2/domain.ts
/** Public contracts for the evolved job service. */

export const ApiVersion = {
  v1: 1,
  v2: 2,
} as const;
export type ApiVersion = (typeof ApiVersion)[keyof typeof ApiVersion];

export const Priority = {
  low: "low",
  normal: "normal",
  high: "high",
} as const;
export type Priority = (typeof Priority)[keyof typeof Priority];

export function parsePriority(value: string): Priority | undefined {
  return Object.values(Priority).find((priority) => priority === value);
}

export type CreateJobRequest = Readonly<{
  jobId: string;
  task: string;
  payload: Uint8Array;
  priority: Priority | undefined;
}>;

export type Job = Readonly<{
  jobId: string;
  task: string;
  payload: Uint8Array;
  priority: Priority;
}>;

export type ValidationField = "job_id" | "task" | "payload" | "priority";
export type ValidationReason =
  | "missing"
  | "invalid_format"
  | "too_long"
  | "control_character"
  | "payload_too_large";

export type ValidationIssue = Readonly<{
  field: ValidationField;
  reason: ValidationReason;
}>;

export type CreateJobResult =
  | Readonly<{ kind: "created"; job: Job }>
  | Readonly<{ kind: "invalid"; issues: readonly ValidationIssue[] }>
  | Readonly<{ kind: "already_exists"; jobId: string }>
  | Readonly<{ kind: "persistence_unavailable" }>;

export const InsertOutcome = {
  inserted: "inserted",
  duplicate: "duplicate",
  unavailableBeforeCommit: "unavailable_before_commit",
} as const;
export type InsertOutcome = (typeof InsertOutcome)[keyof typeof InsertOutcome];

export type JobStore = {
  /** The sole external capability available to the handler. */
  insertIfAbsent(job: Job): InsertOutcome;
};

File: benchmarks/baselines/python/v2/domain.py
"""Public contracts for the evolved job service."""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum, IntEnum, StrEnum
from typing import Protocol


class ApiVersion(IntEnum):
    V1 = 1
    V2 = 2


class Priority(StrEnum):
    LOW = "low"
    NORMAL = "normal"
    HIGH = "high"


@dataclass(frozen=True, slots=True)
class CreateJobRequest:
    job_id: str
    task: str
    payload: bytes
    priority: Priority | None


@dataclass(frozen=True, slots=True)
class Job:
    job_id: str
    task: str
    payload: bytes
    priority: Priority


class ValidationField(StrEnum):
    JOB_ID = "job_id"
    TASK = "task"
    PAYLOAD = "payload"
    PRIORITY = "priority"


class ValidationReason(StrEnum):
    MISSING = "missing"
    INVALID_FORMAT = "invalid_format"
    TOO_LONG = "too_long"
    CONTROL_CHARACTER = "control_character"
    PAYLOAD_TOO_LARGE = "payload_too_large"


@dataclass(frozen=True, slots=True)
class ValidationIssue:
    field: ValidationField
    reason: ValidationReason


@dataclass(frozen=True, slots=True)
class Created:
    job: Job


@dataclass(frozen=True, slots=True)
class Invalid:
    issues: tuple[ValidationIssue, ...]


@dataclass(frozen=True, slots=True)
class AlreadyExists:
    job_id: str


@dataclass(frozen=True, slots=True)
class PersistenceUnavailable:
    pass


type CreateJobResult = Created | Invalid | AlreadyExists | PersistenceUnavailable


class InsertOutcome(Enum):
    INSERTED = "inserted"
    DUPLICATE = "duplicate"
    UNAVAILABLE_BEFORE_COMMIT = "unavailable_before_commit"


class JobStore(Protocol):
    """The sole external capability available to the handler."""

    def insert_if_absent(self, job: Job) -> InsertOutcome:
        """Conditionally persist one job and return the supplied outcome."""
